### PR TITLE
Add XcodeGen support #714

### PR DESCRIPTION
**Pull request by Saik0s**
This PR introduces XcodeGen support #714. The project file is now in `.gitignore` and requires [XcodeGen](https://github.com/yonaskolb/XcodeGen) to generate it from the `project.yml` file. There is also a new `Configs` folder, which contains all the build settings from the old project file. I decided to leave it like that and include `xcconfig` files in `project.yml` to make it cleaner.

The workspace file is still in the repository, but I think it can be completely removed.
Also, I think that instruction how to generate project file should be added somewhere, but I don't know if it should be in `readme.md` or `contributing.md`

## Checklist

- [x] I checked the [**Contributing Guidelines**](https://github.com/SwifterSwift/SwifterSwift/blob/master/CONTRIBUTING.md) before creating this request.
- [ ] New extensions are written in Swift 5.0.
- [ ] New extensions support iOS 8.0+ / tvOS 9.0+ / macOS 10.10+ / watchOS 2.0+, or use `@available` if not.
- [ ] I have added tests for new extensions, and they passed.
- [ ] All extensions have a **clear** comments explaining their functionality, all parameters and return type in English.
- [ ] All extensions are declared as **public**.
- [x] I have added a [changelog](https://github.com/SwifterSwift/SwifterSwift/blob/master/CHANGELOG_GUIDELINES.md) entry describing my changes.

### CODE REVIEW

File: project.yml [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+          - "SwifterSwift/UIKit/UIRefreshControlExtensions.swift"
+          - "SwifterSwift/UIKit/UIActivityExtensions.swift"
+          - "SwifterSwift/CoreAnimation/CAGradientLayerExtensions.swift"
+          - "SwifterSwift/AppKit/*"
+  SwifterSwift-watchOS:
+    type: framework
+    platform: [watchOS]
+    platformSuffix: ""
+    configFiles:
+      Debug: Configs/SwifterSwift-watchOS-Debug.xcconfig
```
Comment (omaralbeik) [inline project.yml]: As much as I love seperating configs into their own files, I don't think we need them for this project
Must of the configs we have are Xcode's defaults

**Comment by LucianoPAlmeida** [comment]
Nice work here, just raising a point that if the users who use Carthage will continue to be able to install, I remember it used to rely on `xcodeproj` to be able to work. If we drop the xcodeproj that may be a problem ... So we probably will have to keep the xcodeproj versioned.

**Comment by omaralbeik** [comment]
@LucianoPAlmeida I remember Xcodegen supports Carthage, users should not face any problems installing the project

**Comment by Saik0s** [comment]
@omaralbeik Unfortunately, I think that @LucianoPAlmeida is right. XcodeGen supports Carthage in terms of knowing the correct path for dependencies installed using Carthage. But in the SwifterSwift case, the repository itself should be supported by Carthage, and to do that, it should have a project file with shared schemes in the root folder of a repository.

**Comment by omaralbeik** [comment]
@Saik0s I thought of adding XcodeGen to ignore project files, in this case, I'm sorry I don't think we will be able to use it 😢

**Comment by omaralbeik** [comment]
I'm very sorry I missed this while opening the issue 😬

**Comment by Saik0s** [comment]
It looks like Carthage is not going even to support SPM in the nearest future because of some [edge cases](https://github.com/Carthage/Carthage/pull/1945). I should have checked it before starting this PR. At least learned something new :)

**Pull request closed without merging**